src/commonMain/kotlin/fr/acinq/lightning/serialization/v2/ChannelState.kt

```diff
@@ -0,0 +1,893 @@
+            is fr.acinq.lightning.transactions.OutgoingHtlc -> OutgoingHtlc(input.add)
+        }
+    }
+}
+
+@Serializable
+data class IncomingHtlc(override val add: UpdateAddHtlc) : DirectedHtlc()
+
+@Serializable
+data class OutgoingHtlc(override val add: UpdateAddHtlc) : DirectedHtlc()
+
+@Serializable
+data class CommitmentSpec(
+    val htlcs: Set<DirectedHtlc>,
+    val feerate: FeeratePerKw,
+    val toLocal: MilliSatoshi,
+    val toRemote: MilliSatoshi
+) {
+    constructor(from: fr.acinq.lightning.transactions.CommitmentSpec) : this(from.htlcs.map { DirectedHtlc.from(it) }.toSet(), from.feerate, from.toLocal, from.toRemote)
+
+    fun export() = fr.acinq.lightning.transactions.CommitmentSpec(htlcs.map { it.to() }.toSet(), feerate, toLocal, toRemote)
+
+}
+
+@Serializable
+data class LocalChanges(val proposed: List<UpdateMessage>, val signed: List<UpdateMessage>, val acked: List<UpdateMessage>) {
+    constructor(from: fr.acinq.lightning.channel.LocalChanges) : this(from.proposed, from.signed, from.acked)
+
+    fun export() = fr.acinq.lightning.channel.LocalChanges(proposed, signed, acked)
+}
+
+@Serializable
+data class RemoteChanges(val proposed: List<UpdateMessage>, val acked: List<UpdateMessage>, val signed: List<UpdateMessage>) {
+    constructor(from: fr.acinq.lightning.channel.RemoteChanges) : this(from.proposed, from.signed, from.acked)
+
+    fun export() = fr.acinq.lightning.channel.RemoteChanges(proposed, signed, acked)
+}
+
+@Serializable
+data class HtlcTxAndSigs(
+    val txinfo: Transactions.TransactionWithInputInfo.HtlcTx,
+    @Serializable(with = ByteVector64KSerializer::class) val localSig: ByteVector64,
+    @Serializable(with = ByteVector64KSerializer::class) val remoteSig: ByteVector64
+) {
+    constructor(from: fr.acinq.lightning.channel.HtlcTxAndSigs) : this(from.txinfo, from.localSig, from.remoteSig)
+
+    fun export() = fr.acinq.lightning.channel.HtlcTxAndSigs(txinfo, localSig, remoteSig)
+}
+
+@Serializable
+data class PublishableTxs(val commitTx: Transactions.TransactionWithInputInfo.CommitTx, val htlcTxsAndSigs: List<HtlcTxAndSigs>) {
+    constructor(from: fr.acinq.lightning.channel.PublishableTxs) : this(from.commitTx, from.htlcTxsAndSigs.map { HtlcTxAndSigs(it) })
+
+    fun export() = fr.acinq.lightning.channel.PublishableTxs(commitTx, htlcTxsAndSigs.map { it.export() })
+}
+
+@Serializable
+data class LocalCommit(val index: Long, val spec: CommitmentSpec, val publishableTxs: PublishableTxs) {
+    constructor(from: fr.acinq.lightning.channel.LocalCommit) : this(from.index, CommitmentSpec(from.spec), PublishableTxs(from.publishableTxs))
+
+    fun export() = fr.acinq.lightning.channel.LocalCommit(index, spec.export(), publishableTxs.export())
+}
+
+@Serializable
+data class RemoteCommit(val index: Long, val spec: CommitmentSpec, @Serializable(with = ByteVector32KSerializer::class) val txid: ByteVector32, @Serializable(with = PublicKeyKSerializer::class) val remotePerCommitmentPoint: PublicKey) {
```

Comment:
Here as well, and in other places in this file, shouldn't we use `Contextual`?

Reply:
same as above